src/tqec/circuit/detector/detector_util.py

```diff
@@ -0,0 +1,406 @@
+    if not any(stim.gate_data(inst.name).produces_measurements for inst in moment):
+        return False
+    if not check_are_all_measurements:
+        return True
+    for inst in moment:
+        if stim.gate_data(inst.name).is_noisy_gate or inst.name in ANNOTATIONS:
+            continue
+        if not stim.gate_data(inst.name).produces_measurements:
+            raise TQECException(
+                f"The measurement moment contains non-measurement instruction: {inst}."
+            )
+    return True
+
+
+def has_reset(moment: stim.Circuit, check_are_all_resets: bool = False) -> bool:
+    """Check if a `stim.Circuit` moment contains reset instructions.
+
+    Args:
+        moment: The moment to check.
+        check_are_all_resets: If `True`, when the moment contains a reset, then all
+            instructions except for annotations and noise instructions must be reset
+            instructions, otherwise exception is raised.
+    """
+    if not any(stim.gate_data(inst.name).is_reset for inst in moment):
+        return False
+    if not check_are_all_resets:
+        return True
+    for inst in moment:
+        # note that measurements are noisy gates by define in stim
+        if (
+            stim.gate_data(inst.name).is_noisy_gate
+            and not stim.gate_data(inst.name).produces_measurements
+            or inst.name in ANNOTATIONS
+        ):
+            continue
+        if not stim.gate_data(inst.name).is_reset:
+            raise TQECException(
+                f"The reset moment contains non-reset instruction: {inst}."
+            )
+    return True
+
+
+def collapse_pauli_strings_at_moment(moment: stim.Circuit, is_reset: bool):
+    def predicate(inst: stim.CircuitInstruction) -> bool:
+        if is_reset:
+            return stim.gate_data(inst.name).is_reset
+        return stim.gate_data(inst.name).produces_measurements
+
+    return [
+        pauli_string
+        for inst in moment
+        if predicate(inst)
+        for pauli_string in collapsing_inst_to_pauli_strings(inst)
+    ]
+
+
+class Fragment:
+    def __init__(
+        self,
+        circuit: stim.Circuit,
+        end_stabilizer_sources: list[PauliString] = (),
+        begin_stabilizer_sources: list[PauliString] = (),
+        sources_for_next_fragment: list[PauliString] = (),
+    ):
+        """A sub-circuit guaranteed to span the locations between two nearest collapsing
+        moments or those outside the error detection regions."""
+
+        self.circuit = circuit
+        self.end_stabilizer_sources = list(end_stabilizer_sources)
+        self.begin_stabilizer_sources = list(begin_stabilizer_sources)
+        self.sources_for_next_fragment = list(sources_for_next_fragment)
+
+    @property
+    def have_detector_sources(self) -> bool:
+        return not (
+            len(self.end_stabilizer_sources) == 0
+            and len(self.begin_stabilizer_sources) == 0
+            and len(self.sources_for_next_fragment) == 0
+        )
```

Comment:
From what I have gathered, `end_stabilizer_sources` stores the `PauliString` instances representing the `reset` collapsing operations and `start_stabilizer_sources` stores the same thing, but for `measurement` collapsing operations right? If that is the case, I really do not understand the `start`/`end` naming, to the point where I think they should even be inverted as, in my head, a `reset` is starting a new stabilizer and a measurement is ending one.

Could you explain the rationale behind that naming?

Reply:
> From what I have gathered, `end_stabilizer_sources` stores the `PauliString` instances representing the `reset` collapsing operations and `start_stabilizer_sources` stores the same thing, but for `measurement` collapsing operations right? 

Yes, it is.

> If that is the case, I really do not understand the start/end naming.

It is indeed a little confusing. I followed the convention/terminology from Austin's proposal. An `end_stabilizer` refers to the stabilizer at the end of a circuit fragment propagated from the start of the fragment. In particular, it is the stabilizer introduced by one/several reset(s) at the beginning of the fragment, passing through the circuit till the end. So the 'end' means the stabilizer is at the end of a fragment, which explains the `end_stabilizer_sources` are resets.

Reply:
Ok, I got it. I am fine with that naming then, but I think a comment to explain would be welcomed. I am working on adding comments and documentation right now, so you do not need to do it. 

I will do a draft PR on your branch shortly to show what I added for the moment.